crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1058,6 +1049,14 @@ impl<'db> Type<'db> {
             _ => KnownClass::Str.to_instance(db),
         }
     }
+
+    #[must_use]
+    pub fn intersect(&self, db: &'db dyn Db, other: Type<'db>) -> Type<'db> {
+        IntersectionBuilder::new(db)
+            .add_positive(*self)
+            .add_positive(other)
+            .build()
+    }
```

Comment:
You might be thinking: isn't this the exact same thing as before? It's not though. We now call `.build()` for every single constraint. Before:
```rs
IntersectionBuilder::new(…)
  .add_positive(bindings_ty)
  .add_positive(constraint_1)
  .add_positive(constraint_2)
  .build();
```
After:
```rs
IntersectionBuilder::new(…)
  .add_positive(
    IntersectionBuilder::new(…)
      .add_positive(bindings_ty)
      .add_positive(constraint_1)
      .build()
  ).add_positive(constraint_2)
  .build()
```

Reply:
Hmm... I'm not sure repeatedly calling `build()` is optimal here. `IntersectionBuilder::build()` calls `InnerIntersectionBuilder::build`, and `InnerIntersectionBuilder::build` has these `shrink_to_fit()` calls. I think it only makes sense to do those when the intersection is "finished", rather than every time we add a positive or negative element to the intersection:

https://github.com/astral-sh/ruff/blob/7dbd8f0f8e725d05cc1d15e6ef1c6cc2ed2b2090/crates/red_knot_python_semantic/src/types/builder.rs#L357-L358

What if we moved the simplifications done by `build()` out into a common method that is called by `build`, `add_positive` and `add_negative`? Would that remove the need to call `build()` every time we added a new element to the intersection?